### ISSUE TITLE
Add functionality for generated maps

### ISSUE THREAD

**Issue by TheRamenChef** [gurkenlabs/litiengine#139]
LITIEngine currently has very little support for maps generated at runtime. I would like to have this functionality for the game I’m working on.

**Comment by steffen-wilke** [comment]
In general, I think this is a good idea. @TheRamenChef could you describe your requirements in more detail? What are the required engine extensions in you eyes?

**Comment by TheRamenChef** [comment]
Basically, I want to be able to make procedurally-generated levels for better replay value.

**Comment by TheRamenChef** [comment]
Any updates on this feature?

**Comment by steffen-wilke** [comment]
In general, I think most of the API in the `de.gurkenlabs.litiengine.tilemap` and `de.gurkenlabs.litiengine.tilemap.xml` namespaces should already allow for generating maps from code. We've done this in a few games we made in the past.

However, I still think there might be some missing or non-public APIs that we didn't come across yet. I'll investigate this and check for any missing pieces.

**Comment by steffen-wilke** [comment]
I just wrote a quick test that checks for the available methods of the XML implementation compared to what is available on the interfaces (public API).

de.gurkenlabs.litiengine.environment.tilemap.xml.CustomProperty	 ✔
de.gurkenlabs.litiengine.environment.tilemap.xml.CustomPropertyProvider	 ✔
de.gurkenlabs.litiengine.environment.tilemap.xml.Frame	 ✔
de.gurkenlabs.litiengine.environment.tilemap.xml.package-info	 ✔
de.gurkenlabs.litiengine.environment.tilemap.xml.PolyShape	 ✔
de.gurkenlabs.litiengine.environment.tilemap.xml.Terrain	 ✔
de.gurkenlabs.litiengine.environment.tilemap.xml.TileAnimation	 ✔
de.gurkenlabs.litiengine.environment.tilemap.xml.TileOffset	 ✔
de.gurkenlabs.litiengine.environment.tilemap.xml.TilesetEntry	 ✔

de.gurkenlabs.litiengine.environment.tilemap.xml.Blueprint
  * build missing
  * keepIds missing
  * getItems missing
  * getBounds missing
  * setPolyline missing

de.gurkenlabs.litiengine.environment.tilemap.xml.GroupLayer
  * setRenderType missing
  * setOpacity missing

de.gurkenlabs.litiengine.environment.tilemap.xml.ImageLayer
  * setRenderType missing
  * setOpacity missing

de.gurkenlabs.litiengine.environment.tilemap.xml.Layer
  * setRenderType missing
  * setOpacity missing

de.gurkenlabs.litiengine.environment.tilemap.xml.MPublic APImage
  * getHeight missing
  * getWidth missing

de.gurkenlabs.litiengine.environment.tilemap.xml.MapObject
  * getBounds missing
  * setPolyline missing

de.gurkenlabs.litiengine.environment.tilemap.xml.MapObjectLayer
  * setRenderType missing
  * setOpacity missing

de.gurkenlabs.litiengine.environment.tilemap.xml.Text
  * getFontName missing
  * getPixelSize missing
  * isBold missing
  * useKerning missing
  * isItalic missing
  * isStrikeout missing
  * isUnderlined missing

de.gurkenlabs.litiengine.environment.tilemap.xml.Tile
  * setTileCoordinate missing

de.gurkenlabs.litiengine.environment.tilemap.xml.TileChunk
  * getValue missing
  * getHeight missing
  * getX missing
  * getWidth missing
  * getY missing

de.gurkenlabs.litiengine.environment.tilemap.xml.TileData$Compression
  * isValid missing

de.gurkenlabs.litiengine.environment.tilemap.xml.TileData$Encoding
  * isValid missing

de.gurkenlabs.litiengine.environment.tilemap.xml.TileData
  * encode missing
  * getValue missing
  * setValue missing
  * getEncoding missing
  * setEncoding missing
  * setCompression missing
  * getCompression missing
  * getTiles missing

de.gurkenlabs.litiengine.environment.tilemap.xml.TileLayer
  * setRenderType missing
  * setOpacity missing

de.gurkenlabs.litiengine.environment.tilemap.xml.Tileset
  * load missing
  * finish missing
  * saveSource missing
  * isExternal missing
  * updateTileTerrain missing

de.gurkenlabs.litiengine.environment.tilemap.xml.TmxMap
  * setHeight missing
  * setWidth missing
  * finish missing
  * setPath missing
  * setVersion missing
  * setTileWidth missing
  * setStaggerIndex missing
  * setOrientation missing
  * setStaggerAxis missing
  * setHexSideLength missing
  * setTiledVersion missing
  * setTileHeight missing
  * setRenderOrder missing
  * getExternalTilesets missing

As a first step, we should iterate over these missing methods and check which ones should be included in the public interface. Then we can define further API that simplifies map generation, e.g. a helper class that simplifies the generation (also see #329 and [This forum post](https://forum.litiengine.com/d/116-setting-tiles-in-maps-on-the-fly-pull-request/6))

**Comment by evnsch** [comment]
Not to be greedy, but can you point a link to these past games? 
Litiengine doesn't seem to be automagically reading & using collision boxes from tmx files.
It would be helpful to see how you get that to work

**Comment by steffen-wilke** [comment]
> Not to be greedy, but can you point a link to these past games?

Sorry, these were internal projects. But I'm currently working on extending [Gurk Nukem](https://github.com/gurkenlabs/litiengine-gurk-nukem) with a mechanism to generate maps at runtime as an example project.

In general: collision boxes are just `MapObjects` with the type `MapObjectType.COLLISIONBOX`.
Our utiliti editor is basically the best publically available example for adding map objects to a map.
Basically, what you need is a `IMapObjectLayer` and then you can add as many objects of different types as you desire:

https://github.com/gurkenlabs/litiengine/blob/3455fa6b1b98ccc4cd8e6fbd15ed34f273776937/utiliti/src/de/gurkenlabs/utiliti/components/MapComponent.java#L378-L389

https://github.com/gurkenlabs/litiengine/blob/3455fa6b1b98ccc4cd8e6fbd15ed34f273776937/utiliti/src/de/gurkenlabs/utiliti/components/MapComponent.java#L709-L716

**Comment by evnsch** [comment]
Genius, that will work in theory. I'll check it out

**Comment by steffen-wilke** [comment]
96be92d29de0299db289e0a0b9e3e91db5cd6209 introduces a new API to support generating maps: `Resources.map.generate(...)`

**Usage**
```java
IMap generatedMap;
try (MapGenerator generator = Resources.maps().generate(MapOrientations.ORTHOGONAL, "mymap", 100, 100, 16, 16, Resources.tilesets().get("tileset.tsx"))) {
  tileLayer = generator.addTileLayer(RenderType.GROUND, (x, y) -> {
    if (x == y) {
      // draw a diagonal in another tile color
      return 2;
    }

    // fill the entire map with this tile
    return 1;
  });

  // set an explicit tile at a location
  tileLayer.setTile(10, 10, 3);

  // add a collision box to the map
  generator.add(new CollisionBox(0, 50, 100, 10));

  generatedMap = generator.getMap();
}
```
